README.md
# **videovigilancia-aumentada**

<!-- TABLE OF CONTENTS -->
<details>
  <summary>Tabla de contenidos</summary>
  <ol>
    <li>
      <a href="#Sobre-el-proyecto">Sobre el proyecto</a>
    </li>
    <li>
      <a href="#tecnologias-e-instalacion">Tecnologías e instalación</a>
      <ul>
        <li><a href="#prerequisitos">Prerequisitos</a></li>
        <li><a href="#instalacion">Instalacion</a></li>
      </ul>
    </li>
    <li><a href="#forma-de-uso">Forma de uso</a></li>
    <li><a href="#roadmap">Roadmap</a></li>
  </ol>
</details>

<!-- ABOUT THE PROJECT -->
## Sobre el proyecto
<p id="Sobre-el-proyecto">Repositorio de librerías y demos para la videovigilancia aumentada basada en Inteligencia Artificial orientada a la detección de personas, clasificación de comportamientos y reconocimiento facial.</p>

## Tecnologías e instalación
<p id="tecnologias-e-instalacion">A continuación se listan las tecnologías usadas para la elaboración de este proyecto, así como también la forma de instalar el mismo para correr las demos.</p>

### Prerequisitos
<p id="prerequisitos">Este proyecto es un aplicativo web que usa las siguinetes tecnologías:</p>

* **Node.js** para back-end, 
* **MongoDB** para las bases de datos.
* **Pytorch** para el modelo de reconocimiento de poses.
* ... TODO

### Instalación
<p id="instalacion">Seguir los siguientes pasos para echar a andar la demo.</p>
TODO

## Forma de uso
<p id="forma-de-uso">Para correr el proyecto en un servidor local seguir los siguientes pasos:</p>
TODO

## Roadmap
<p id="roadmap">El roadmap del proyecto se presenta a continuación.</p>

- [ ] Desarrollo del primer demo usando cloud computing.
- [ ] Desarrollo del demo usando edge computing.
- [ ] ...


CHANGELOG.md
# Change Log

## [1.0.3] 2023-07-27
### Add
- Se agregó el código en python para realizar las predicciones de rostro, pose y acciones de violencia de robo a mano armada.

## [1.0.2] 2023-07-11
### Add
- Se agrego el servidor creado en nodeJs para comunicarse con el app en android y con el servidor de las camaras con IA

## [1.0.1] 2023-08-09
### Add
- Se agrego el archivo .sh para el despliegue de una instancia EC2 de AWS.

## [1.0.0] 2023-07-02
### Add
- [VI-12](https://tekvot.atlassian.net/browse/VI-12) Incluir el Readme.md para detallar aspectos del proyecto y agregar el changelog, para llevar un orden de las actualizaciones que se realicen en este repo.


server-nodejs/index.js
const { Server } = require('socket.io')
const { log } = require('console');
const express = require('express');
const http = require('http');
const createImage = require('./bringImage');

//Import middlewares
const cors = require('cors');
const bodyParser = require('body-parser');



//Create a express server   
const app = express();
const server = http.createServer(app);
const port = 4000

//Middlewares
app.use(bodyParser.json());
app.use(cors());

//Socket.io
const io = new Server({
    //acept all cors origin
    cors: {
        origin: "*"
    }
});
//io port
io.listen(5000);



//Routes API
app.get('/', (req, res) => {
    res.send('Hello World!')
});


app.post('/', (req, res) => {
    if (req.body.dni && req.body.name && req.body.image64 && req.body.alert) {
        console.log("Hay una alerta y estos son los datos:", req.body);
        const newBody = {
            dni: req.body.dni,
            name: req.body.name,
            image64: req.body.image64,
            alert: req.body.alert
        }
        console.log('🚀 ~ file: index.js:49 ~ newBody:', newBody)
        io.emit('envioAlApp', newBody);
        res.send(newBody)
    } else if (req.body.alert === false) {
        console.log("No hay alerta");
        io.emit('envioAlApp', req.body);
        res.send(req.body)
    }
});



//Socket connection
io.on('connection', (socket) => {
    console.log('a user connected');
});


server.listen(port, () => {
    console.log(`CAMIA Backend en http://localhost:${port}`)
});


//Modelo
// {
//     "dni":dni,
//     "name":name,
//     "image64":cadena,
//     "alert":alert
// }

python-scripts/src/body.py
import cv2
import numpy as np
import math
import time
from scipy.ndimage.filters import gaussian_filter
import matplotlib.pyplot as plt
import matplotlib
import torch
from torchvision import transforms
#import caffemodel2pytorch
from src import util
from src.model import bodypose_model

class Body(object):
    angle_degrees1=0.0
    angle_degrees2=0.0
    def __init__(self, model_path):
        if model_path.split('.')[1] != "prototxt": 
            self.model = bodypose_model()
            if torch.cuda.is_available():
                self.model = self.model.cuda()
            model_dict = util.transfer(self.model, torch.load(model_path))
            self.model.load_state_dict(model_dict)
            self.model.eval()
        """
        else:
            self.model = caffemodel2pytorch.Net(
                prototxt = 'models/'+model_path,
                weights = 'models/'+model_path.split('.')[0]+".caffemodel"
                #caffe_proto = 'https://raw.githubusercontent.com/BVLC/caffe/master/src/caffe/proto/caffe.proto'
            )
            self.model.cuda()
            self.model.eval()
            torch.set_grad_enabled(False)
        """

    def __call__(self, oriImg):
        # scale_search = [0.5, 1.0, 1.5, 2.0]
        scale_search = [0.5]
        boxsize = 368
        stride = 8
        padValue = 128
        thre1 = 0.1
        thre2 = 0.05
        multiplier = [x * boxsize / oriImg.shape[0] for x in scale_search]
        heatmap_avg = np.zeros((oriImg.shape[0], oriImg.shape[1], 19))
        paf_avg = np.zeros((oriImg.shape[0], oriImg.shape[1], 38))

        for m in range(len(multiplier)):
            scale = multiplier[m]
            imageToTest = cv2.resize(oriImg, (0, 0), fx=scale, fy=scale, interpolation=cv2.INTER_CUBIC)
            imageToTest_padded, pad = util.padRightDownCorner(imageToTest, stride, padValue)
            im = np.transpose(np.float32(imageToTest_padded[:, :, :, np.newaxis]), (3, 2, 0, 1)) / 256 - 0.5
            im = np.ascontiguousarray(im)

            data = torch.from_numpy(im).float()
            if torch.cuda.is_available():
                data = data.cuda()
            # data = data.permute([2, 0, 1]).unsqueeze(0).float()
            with torch.no_grad():
                Mconv7_stage6_L1, Mconv7_stage6_L2 = self.model(data)
            Mconv7_stage6_L1 = Mconv7_stage6_L1.cpu().numpy()
            Mconv7_stage6_L2 = Mconv7_stage6_L2.cpu().numpy()

            # extract outputs, resize, and remove padding
            # heatmap = np.transpose(np.squeeze(net.blobs[output_blobs.keys()[1]].data), (1, 2, 0))  # output 1 is heatmaps
            heatmap = np.transpose(np.squeeze(Mconv7_stage6_L2), (1, 2, 0))  # output 1 is heatmaps
            heatmap = cv2.resize(heatmap, (0, 0), fx=stride, fy=stride, interpolation=cv2.INTER_CUBIC)
            heatmap = heatmap[:imageToTest_padded.shape[0] - pad[2], :imageToTest_padded.shape[1] - pad[3], :]
            heatmap = cv2.resize(heatmap, (oriImg.shape[1], oriImg.shape[0]), interpolation=cv2.INTER_CUBIC)

            # paf = np.transpose(np.squeeze(net.blobs[output_blobs.keys()[0]].data), (1, 2, 0))  # output 0 is PAFs
            paf = np.transpose(np.squeeze(Mconv7_stage6_L1), (1, 2, 0))  # output 0 is PAFs
            paf = cv2.resize(paf, (0, 0), fx=stride, fy=stride, interpolation=cv2.INTER_CUBIC)
            paf = paf[:imageToTest_padded.shape[0] - pad[2], :imageToTest_padded.shape[1] - pad[3], :]
            paf = cv2.resize(paf, (oriImg.shape[1], oriImg.shape[0]), interpolation=cv2.INTER_CUBIC)

            heatmap_avg += heatmap_avg + heatmap / len(multiplier)
            paf_avg += + paf / len(multiplier)

        all_peaks = []
        peak_counter = 0

        for part in range(18):
            map_ori = heatmap_avg[:, :, part]
            one_heatmap = gaussian_filter(map_ori, sigma=3)

            map_left = np.zeros(one_heatmap.shape)
            map_left[1:, :] = one_heatmap[:-1, :]
            map_right = np.zeros(one_heatmap.shape)
            map_right[:-1, :] = one_heatmap[1:, :]
            map_up = np.zeros(one_heatmap.shape)
            map_up[:, 1:] = one_heatmap[:, :-1]
            map_down = np.zeros(one_heatmap.shape)
            map_down[:, :-1] = one_heatmap[:, 1:]

            peaks_binary = np.logical_and.reduce(
                (one_heatmap >= map_left, one_heatmap >= map_right, one_heatmap >= map_up, one_heatmap >= map_down, one_heatmap > thre1))
            peaks = list(zip(np.nonzero(peaks_binary)[1], np.nonzero(peaks_binary)[0]))  # note reverse
            peaks_with_score = [x + (map_ori[x[1], x[0]],) for x in peaks]
            peak_id = range(peak_counter, peak_counter + len(peaks))
            peaks_with_score_and_id = [peaks_with_score[i] + (peak_id[i],) for i in range(len(peak_id))]

            all_peaks.append(peaks_with_score_and_id)
            peak_counter += len(peaks)

        # find connection in the specified sequence, center 29 is in the position 15
        limbSeq = [[2, 3], [2, 6], [3, 4], [4, 5], [6, 7], [7, 8], [2, 9], [9, 10], \
                   [10, 11], [2, 12], [12, 13], [13, 14], [2, 1], [1, 15], [15, 17], \
                   [1, 16], [16, 18], [3, 17], [6, 18]]
        # the middle joints heatmap correpondence
        mapIdx = [[31, 32], [39, 40], [33, 34], [35, 36], [41, 42], [43, 44], [19, 20], [21, 22], \
                  [23, 24], [25, 26], [27, 28], [29, 30], [47, 48], [49, 50], [53, 54], [51, 52], \
                  [55, 56], [37, 38], [45, 46]]

        connection_all = []
        special_k = []
        mid_num = 10

        for k in range(len(mapIdx)):
            score_mid = paf_avg[:, :, [x - 19 for x in mapIdx[k]]]
            candA = all_peaks[limbSeq[k][0] - 1]
            candB = all_peaks[limbSeq[k][1] - 1]
            nA = len(candA)
            nB = len(candB)
            indexA, indexB = limbSeq[k]
            if (nA != 0 and nB != 0):
                connection_candidate = []
                for i in range(nA):
                    for j in range(nB):
                        vec = np.subtract(candB[j][:2], candA[i][:2])
                        norm = math.sqrt(vec[0] * vec[0] + vec[1] * vec[1])
                        norm = max(0.001, norm)
                        vec = np.divide(vec, norm)

                        startend = list(zip(np.linspace(candA[i][0], candB[j][0], num=mid_num), \
                                            np.linspace(candA[i][1], candB[j][1], num=mid_num)))

                        vec_x = np.array([score_mid[int(round(startend[I][1])), int(round(startend[I][0])), 0] \
                                          for I in range(len(startend))])
                        vec_y = np.array([score_mid[int(round(startend[I][1])), int(round(startend[I][0])), 1] \
                                          for I in range(len(startend))])

                        score_midpts = np.multiply(vec_x, vec[0]) + np.multiply(vec_y, vec[1])
                        score_with_dist_prior = sum(score_midpts) / len(score_midpts) + min(
                            0.5 * oriImg.shape[0] / norm - 1, 0)
                        criterion1 = len(np.nonzero(score_midpts > thre2)[0]) > 0.8 * len(score_midpts)
                        criterion2 = score_with_dist_prior > 0
                        if criterion1 and criterion2:
                            connection_candidate.append(
                                [i, j, score_with_dist_prior, score_with_dist_prior + candA[i][2] + candB[j][2]])

                connection_candidate = sorted(connection_candidate, key=lambda x: x[2], reverse=True)
                connection = np.zeros((0, 5))
                for c in range(len(connection_candidate)):
                    i, j, s = connection_candidate[c][0:3]
                    if (i not in connection[:, 3] and j not in connection[:, 4]):
                        connection = np.vstack([connection, [candA[i][3], candB[j][3], s, i, j]])
                        if (len(connection) >= min(nA, nB)):
                            break

                connection_all.append(connection)
            else:
                special_k.append(k)
                connection_all.append([])

        # last number in each row is the total parts number of that person
        # the second last number in each row is the score of the overall configuration
        subset = -1 * np.ones((0, 20))
        candidate = np.array([item for sublist in all_peaks for item in sublist])

        for k in range(len(mapIdx)):
            if k not in special_k:
                partAs = connection_all[k][:, 0]
                partBs = connection_all[k][:, 1]
                indexA, indexB = np.array(limbSeq[k]) - 1

                for i in range(len(connection_all[k])):  # = 1:size(temp,1)
                    found = 0
                    subset_idx = [-1, -1]
                    for j in range(len(subset)):  # 1:size(subset,1):
                        if subset[j][indexA] == partAs[i] or subset[j][indexB] == partBs[i]:
                            subset_idx[found] = j
                            found += 1

                    if found == 1:
                        j = subset_idx[0]
                        if subset[j][indexB] != partBs[i]:
                            subset[j][indexB] = partBs[i]
                            subset[j][-1] += 1
                            subset[j][-2] += candidate[partBs[i].astype(int), 2] + connection_all[k][i][2]
                    elif found == 2:  # if found 2 and disjoint, merge them
                        j1, j2 = subset_idx
                        membership = ((subset[j1] >= 0).astype(int) + (subset[j2] >= 0).astype(int))[:-2]
                        if len(np.nonzero(membership == 2)[0]) == 0:  # merge
                            subset[j1][:-2] += (subset[j2][:-2] + 1)
                            subset[j1][-2:] += subset[j2][-2:]
                            subset[j1][-2] += connection_all[k][i][2]
                            subset = np.delete(subset, j2, 0)
                        else:  # as like found == 1
                            subset[j1][indexB] = partBs[i]
                            subset[j1][-1] += 1
                            subset[j1][-2] += candidate[partBs[i].astype(int), 2] + connection_all[k][i][2]

                    # if find no partA in the subset, create a new subset
                    elif not found and k < 17:
                        row = -1 * np.ones(20)
                        row[indexA] = partAs[i]
                        row[indexB] = partBs[i]
                        row[-1] = 2
                        row[-2] = sum(candidate[connection_all[k][i, :2].astype(int), 2]) + connection_all[k][i][2]
                        subset = np.vstack([subset, row])
        # delete some rows of subset which has few parts occur
        deleteIdx = []
        for i in range(len(subset)):
            if subset[i][-1] < 4 or subset[i][-2] / subset[i][-1] < 0.4:
                deleteIdx.append(i)
        subset = np.delete(subset, deleteIdx, axis=0)

        # subset: n*20 array, 0-17 is the index in candidate, 18 is the total score, 19 is the total parts
        # candidate: x, y, score, id
        return candidate, subset
    def calculate_angle(self, candidate, subset):
        right_shoulder = None
        right_elbow = None
        right_wrist = None
        cabeza     = None
        for i in range(len(subset)):
            cabeza_idx = int(subset[i][0])
            right_shoulder_idx = int(subset[i][2])
            right_elbow_idx = int(subset[i][3])
            right_wrist_idx = int(subset[i][4])

            if right_shoulder_idx != -1 and right_elbow_idx != -1 and right_wrist_idx != -1 and cabeza_idx != -1:
                right_shoulder = candidate[right_shoulder_idx][:2]
                right_elbow = candidate[right_elbow_idx][:2]
                right_wrist = candidate[right_wrist_idx][:2]
                cabeza      = candidate[cabeza_idx][:2]
                break

        if right_shoulder is not None and right_elbow is not None and right_wrist is not None and cabeza is not None:
            cabeza_shouler = np.subtract(cabeza,right_shoulder)
            shoulder_to_elbow = np.subtract(right_elbow, right_shoulder)
            elbow_to_wrist = np.subtract(right_wrist, right_elbow)

            cabeza_shouler_norm = np.linalg.norm(cabeza_shouler)
            shoulder_to_elbow_norm = np.linalg.norm(shoulder_to_elbow)
            elbow_to_wrist_norm = np.linalg.norm(elbow_to_wrist)

            if shoulder_to_elbow_norm > 0 and elbow_to_wrist_norm > 0 and cabeza_shouler_norm > 0:
                cabeza_shouler    = cabeza_shouler/cabeza_shouler_norm
                shoulder_to_elbow = shoulder_to_elbow / shoulder_to_elbow_norm
                elbow_to_wrist = elbow_to_wrist / elbow_to_wrist_norm

                angle1 = np.arccos(np.clip(np.dot(shoulder_to_elbow, elbow_to_wrist), -1.0, 1.0))
                angle2 = np.arccos(np.clip(np.dot(cabeza_shouler, shoulder_to_elbow), -1.0, 1.0))
                angle_degrees1 = np.degrees(angle1)
                angle_degrees2 = np.degrees(angle2)

                self.angle_degrees1=angle_degrees1
                self.angle_degrees2=angle_degrees2

if __name__ == "__main__":
    body_estimation = Body('../model/body_pose_model.pth')

    test_image = '../images/ski.jpg'
    oriImg = cv2.imread(test_image)  # B,G,R order
    candidate, subset = body_estimation(oriImg)
    canvas = util.draw_bodypose(oriImg, candidate, subset)
    plt.imshow(canvas[:, :, [2, 1, 0]])
    plt.show()


python-scripts/src/hand.py
import cv2
import json
import numpy as np
import math
import time
from scipy.ndimage.filters import gaussian_filter
import matplotlib.pyplot as plt
import matplotlib
import torch
#import caffemodel2pytorch
from skimage.measure import label

from src.model import handpose_model
from src import util

class Hand(object):
    def __init__(self, model_path):
        if model_path.split('.')[1] != "prototxt": 
            self.model = handpose_model()
            if torch.cuda.is_available():
                self.model = self.model.cuda()
            model_dict = util.transfer(self.model, torch.load(model_path))
            self.model.load_state_dict(model_dict)
            self.model.eval()
        """
        else:
            self.model = caffemodel2pytorch.Net(
                prototxt = 'models/'+model_path,
                weights = 'models/'+model_path.split('.')[0]+".caffemodel"
                #caffe_proto = 'https://raw.githubusercontent.com/BVLC/caffe/master/src/caffe/proto/caffe.proto'
            )
            self.model.cuda()
            self.model.eval()
            torch.set_grad_enabled(False)
        """
    def __call__(self, oriImg):
        scale_search = [0.5, 1.0, 1.5, 2.0]
        # scale_search = [0.5]
        boxsize = 368
        stride = 8
        padValue = 128
        thre = 0.05
        multiplier = [x * boxsize / oriImg.shape[0] for x in scale_search]
        heatmap_avg = np.zeros((oriImg.shape[0], oriImg.shape[1], 22))
        # paf_avg = np.zeros((oriImg.shape[0], oriImg.shape[1], 38))

        for m in range(len(multiplier)):
            scale = multiplier[m]
            imageToTest = cv2.resize(oriImg, (0, 0), fx=scale, fy=scale, interpolation=cv2.INTER_CUBIC)
            imageToTest_padded, pad = util.padRightDownCorner(imageToTest, stride, padValue)
            im = np.transpose(np.float32(imageToTest_padded[:, :, :, np.newaxis]), (3, 2, 0, 1)) / 256 - 0.5
            im = np.ascontiguousarray(im)

            data = torch.from_numpy(im).float()
            if torch.cuda.is_available():
                data = data.cuda()
            # data = data.permute([2, 0, 1]).unsqueeze(0).float()
            with torch.no_grad():
                output = self.model(data).cpu().numpy()
                # output = self.model(data).numpy()q

            # extract outputs, resize, and remove padding
            heatmap = np.transpose(np.squeeze(output), (1, 2, 0))  # output 1 is heatmaps
            heatmap = cv2.resize(heatmap, (0, 0), fx=stride, fy=stride, interpolation=cv2.INTER_CUBIC)
            heatmap = heatmap[:imageToTest_padded.shape[0] - pad[2], :imageToTest_padded.shape[1] - pad[3], :]
            heatmap = cv2.resize(heatmap, (oriImg.shape[1], oriImg.shape[0]), interpolation=cv2.INTER_CUBIC)

            heatmap_avg += heatmap / len(multiplier)

        all_peaks = []
        for part in range(21):
            map_ori = heatmap_avg[:, :, part]
            one_heatmap = gaussian_filter(map_ori, sigma=3)
            binary = np.ascontiguousarray(one_heatmap > thre, dtype=np.uint8)
            # 全部小于阈值
            if np.sum(binary) == 0:
                all_peaks.append([0, 0])
                continue
            label_img, label_numbers = label(binary, return_num=True, connectivity=binary.ndim)
            max_index = np.argmax([np.sum(map_ori[label_img == i]) for i in range(1, label_numbers + 1)]) + 1
            label_img[label_img != max_index] = 0
            map_ori[label_img == 0] = 0

            y, x = util.npmax(map_ori)
            all_peaks.append([x, y])
        return np.array(all_peaks)

if __name__ == "__main__":
    hand_estimation = Hand('../model/hand_pose_model.pth')

    # test_image = '../images/hand.jpg'
    test_image = '../images/hand.jpg'
    oriImg = cv2.imread(test_image)  # B,G,R order
    peaks = hand_estimation(oriImg)
    canvas = util.draw_handpose(oriImg, peaks, True)
    cv2.imshow('', canvas)
    cv2.waitKey(0)

python-scripts/demo_camera.py
import cv2
import copy
import numpy as np
import torch
import mediapipe as mp
import os
#os.environ['OPENCV_FFMPEG_CAPTURE_OPTIONS'] = 'rtsp_transport;udp'
import requests
from src import model
from src import util
from src.body import Body
import base64
import face_recognition
import imutils
from imutils.video import VideoStream
mp_face_detection = mp.solutions.face_detection
face_detection=mp_face_detection.FaceDetection(min_detection_confidence=0.5)
rtsp_url = "rtsp://admin:KCGZBB@192.168.100.224:554/"
def image_to_base64(image):
    # Codifica la imagen en formato base64
    _, buffer = cv2.imencode('.jpg', image)
    encoded_image = base64.b64encode(buffer)
    return encoded_image.decode('utf-8')
font = cv2.FONT_HERSHEY_SIMPLEX
font_scale = 0.5
font_thickness = 1

body_estimation = Body('model/body_pose_model.pth')

def load_faces(folder):
    face_encodings = []
    face_names = []
    for filename in os.listdir(folder):
        if filename.endswith(".jpg") or filename.endswith(".png") or filename.endswith(".JPG") or filename.endswith(".jpeg"):
                #img = cv2.imread(os.path.join(folder+"/"+d, filename))
                #image_rgb = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
                #results = face_detection.process(image_rgb)
            face_image = face_recognition.load_image_file(os.path.join(folder, filename))
            face_locations = face_recognition.face_locations(face_image)
                #bbox=results.detections[0].location_data.relative_bounding_box
                #print(bbox)
                #print(face_locations)
                #image_height, image_width, _ = img.shape
                #bbox_start_point = [int(bbox.xmin * image_width), int(bbox.ymin * image_height), int((bbox.xmin + bbox.width) * image_width), int((bbox.ymin + bbox.height) * image_height)]
            if len(face_locations) != 0:
                aux = face_recognition.face_encodings(face_image, face_locations,model="cnn")[0]
                face_encodings.append(aux)
            face_names.append(filename.split('.')[0])
            
    return face_encodings, face_names

face_encodings, face_names = load_faces("rostros")

print(f"Torch device: {torch.cuda.get_device_name()}")

#cap = cv2.VideoCapture(0) #"rtsp://prel1992:Fores753***@192.168.100.103:554/stream1")
video_stream = VideoStream(rtsp_url).start()

def calculate_angle(frame, p1, p2, p3):
    # Calcula el ángulo entre tres puntos en el plano 2D
    v1 = np.array([p1[0] - p2[0], p1[1] - p2[1]])
    v2 = np.array([p3[0] - p2[0], p3[1] - p2[1]])

    dot_product = np.dot(v1, v2)
    norm_product = np.linalg.norm(v1) * np.linalg.norm(v2)
    cosine_angle = dot_product / norm_product
    angle = np.arccos(cosine_angle) * 180 / np.pi
    cv2.putText(frame, f"Right Angle: {angle:.2f}", (10, 20), font, font_scale, (255, 255, 255), font_thickness)
    return angle

def pose_estimation_logic(frame, candidate, subset,permiso):
    neck_idx = np.where(subset[:, 0] != 1)[0]
    if len(neck_idx) > 0:
        neck_idx = neck_idx[0]
        right_shoulder_idx = subset[neck_idx][2]
        left_shoulder_idx = subset[neck_idx][5]
        cabeza = subset[neck_idx][0]
        brazo_izquierdo = subset[neck_idx][5],subset[neck_idx][6],subset[neck_idx][7]
        brazo_derecho = subset[neck_idx][2],subset[neck_idx][3],subset[neck_idx][4]
        if right_shoulder_idx != -1 and left_shoulder_idx != -1 and cabeza != -1 and brazo_izquierdo[1] != -1 and brazo_izquierdo[2] != -1 and brazo_derecho[1] != -1 and brazo_derecho[2] != -1:
            right_shoulder = candidate[int(right_shoulder_idx)]
            left_shoulder = candidate[int(left_shoulder_idx)]
            shoulder_angle = calculate_angle(frame, right_shoulder, left_shoulder, (frame.shape[1] // 2, frame.shape[0]))
            angulo_codo_cabeza = calculate_angle(frame,brazo_derecho[2],cabeza,(frame.shape[1] // 2, frame.shape[0]))
            print(angulo_codo_cabeza)
            if shoulder_angle > 180 and permiso:
                data = {
                    'name': 'bryan',
                    'dni': 178529634,
                    'image64': image_to_base64(cv2.resize(oriImg,(0,0),fx=0.5,fy=0.5)),
                    'alert':1
                }
                print('envio')
                response = requests.post('http://localhost:4000', json=data)
                permiso=False
    return False
permiso=True
while True:
    #ret, oriImg = cap.read()
    oriImg = video_stream.read()
    oriImg_=cv2.resize(oriImg,(0,0),fx=0.25,fy=0.25)
    candidate, subset = body_estimation(oriImg_)
    #canvas = copy.deepcopy(oriImg)
    #subset = (np.array(subset,dtype=np.uint32)*4).tolist()
    canvas = util.draw_bodypose(oriImg, candidate, subset,4)
    body_estimation.calculate_angle(candidate=candidate,subset=subset)
    #pose_estimation_logic(canvas,candidate, subset,permiso)
    if body_estimation.angle_degrees1 > 90 :
        data = {
                    'name': 'bryan',
                    'dni': 178529634,
                    'image64': image_to_base64(cv2.resize(oriImg,(0,0),fx=0.25,fy=0.25)),
                    'alert':1
                }
        print('envio')
        response = requests.post('http://localhost:4000', json=data)
        permiso=False

    face_locations = face_recognition.face_locations(oriImg)
    face_encodings_ = face_recognition.face_encodings(oriImg, face_locations,model="cnn")

    for face_encoding, face_location in zip(face_encodings_, face_locations):
        matches = face_recognition.compare_faces(face_encodings, face_encoding)
        name = "Unknown"
        if True in matches:
            first_match_index = np.argmax(True)
            if face_names != None:
                if len(face_encoding)!=0 and len(face_encoding)>first_match_index:
                    name = face_names[first_match_index]
        y1, x2, y2, x1 = face_location
        cv2.rectangle(canvas, (4*x1, 4*y1), (4*x2, 4*y2), (0, 0, 255), 3)
        cv2.putText(canvas, name, (4*x1, 4*y1 - 40), cv2.FONT_HERSHEY_SIMPLEX, 0.9, (0, 0, 255), 2)
    #canvas=cv2.resize(canvas,(0,0),fx=1/0.4,fy=1/0.4)
    cv2.imshow('demo', canvas)
    if cv2.waitKey(5) & 0xFF == ord('q'):
        break
#cap.release()
cv2.destroyAllWindows()

python-scripts/clases.py
import cv2
import copy
import numpy as np
import torch
import mediapipe as mp
import os
import requests
from src import model
from src import util
from src.body import Body
import base64
import face_recognition
import imutils
from imutils.video import VideoStream
from threading import Thread,Timer

class ImageUtils:
    mp_face_detection = mp.solutions.face_detection
    face_detection=mp_face_detection.FaceDetection(min_detection_confidence=0.5)
    rtsp_url = "rtsp://prel1992:Fores753***@192.168.100.213:554/stream1"
    font = cv2.FONT_HERSHEY_SIMPLEX
    font_scale = 0.5
    font_thickness = 1
    body_estimation = Body('model/body_pose_model.pth')
    face_encodings = []
    face_names = []
    faces_locations=[]
    video_stream = VideoStream(src=0).start()
    frame = None
    names=[]
    candidate = None
    subset    = None
    canvas    = None
    permiso=True
    draw_pose_permission=False
    draw_face_permission=False
    hilos=[]
    def __init__(self):
        self.load_faces('rostros')
        self.hilos.append(Thread(target=self.loop))
        self.hilos.append(Timer(0.5,self.pose_estimation_logic))
        self.hilos.append(Timer(1,self.face_identification))
        self.hilos.append(Timer(0.2,self.estimation_pose))
        for hilo in self.hilos:
            hilo.start()

    def image_to_base64(self,image):
        _, buffer = cv2.imencode('.jpg', image)
        encoded_image = base64.b64encode(buffer)
        return encoded_image.decode('utf-8')
    
    def load_faces(self,folder):
        for filename in os.listdir(folder):
            if filename.endswith(".jpg") or filename.endswith(".png") or filename.endswith(".JPG") or filename.endswith(".jpeg"):
                    #img = cv2.imread(os.path.join(folder+"/"+d, filename))
                    #image_rgb = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
                    #results = face_detection.process(image_rgb)
                face_image = face_recognition.load_image_file(os.path.join(folder, filename))
                face_locations = face_recognition.face_locations(face_image)
                    #bbox=results.detections[0].location_data.relative_bounding_box
                    #print(bbox)
                    #print(face_locations)
                    #image_height, image_width, _ = img.shape
                    #bbox_start_point = [int(bbox.xmin * image_width), int(bbox.ymin * image_height), int((bbox.xmin + bbox.width) * image_width), int((bbox.ymin + bbox.height) * image_height)]
                if len(face_locations) != 0:
                    aux = face_recognition.face_encodings(face_image, face_locations,model="cnn")[0]
                    self.face_encodings.append(aux)
                self.face_names.append(filename.split('.')[0])

    def calculate_angle(self, p1, p2, p3):
        v1 = np.array([p1[0] - p2[0], p1[1] - p2[1]])
        v2 = np.array([p3[0] - p2[0], p3[1] - p2[1]])
        dot_product = np.dot(v1, v2)
        norm_product = np.linalg.norm(v1) * np.linalg.norm(v2)
        cosine_angle = dot_product / norm_product
        angle = np.arccos(cosine_angle) * 180 / np.pi
        cv2.putText(self.frame, f"Right Angle: {angle:.2f}", (10, 20), self.font, self.font_scale, (255, 255, 255), self.font_thickness)
        return angle
    
    def pose_estimation_logic(self):
        if self.candidate !=None and self.subset != None:
            neck_idx = np.where(self.subset[:, 0] != 1)[0]
            if len(neck_idx) > 0:
                neck_idx = neck_idx[0]
                right_shoulder_idx = self.subset[neck_idx][2]
                left_shoulder_idx = self.subset[neck_idx][5]
                if right_shoulder_idx != -1 and left_shoulder_idx != -1:
                    right_shoulder = self.candidate[int(right_shoulder_idx)]
                    left_shoulder = self.candidate[int(left_shoulder_idx)]
                    shoulder_angle = self.calculate_angle(self.frame, right_shoulder, left_shoulder, (self.frame.shape[1] // 2, self.frame.shape[0]))
                    print(shoulder_angle)
                    if shoulder_angle > 90 and self.permiso:
                        data = {
                            'name': 'bryan',
                            'dni': 70886887,
                            'image64': self.image_to_base64(cv2.resize(self.frame,(0,0),fx=0.5,fy=0.5)),
                            'alert':1
                        }
                        print('envio')
                        response = requests.post('http://localhost:4000', json=data)
                        self.permiso=False
    def estimation_pose(self):
        if self.frame is not None:
            self.candidate, self.subset = self.body_estimation(self.frame)
            #self.canvas = copy.deepcopy(self.frame)
            self.frame = util.draw_bodypose(self.frame, self.candidate, self.subset)
        h=Timer(0.35,self.estimation_pose)
        h.start()
            
    def face_identification(self):
        if self.frame is not None:
            self.names=[]
            face_locations = face_recognition.face_locations(self.frame)
            face_encodings = face_recognition.face_encodings(self.frame, face_locations,model="cnn")
            for face_encoding, face_location in zip(face_encodings, face_locations):
                matches = face_recognition.compare_faces(self.face_encodings, face_encoding)
                if any(matches):
                    first_match_index = np.argmax(True)
                    if self.face_names != None:
                        if len(face_encoding)!=0 and len(face_encoding)>first_match_index:
                            self.names.append(self.face_names[first_match_index])
                        else:
                            self.names.append("Desconocido")
            h=Timer(0.35,self.face_identification)
            h.start()
    def loop(self):
        while True:
            self.frame = self.video_stream.read()
            self.frame = cv2.resize(self.frame,(0,0),fx=0.4,fy=0.4)
            if self.candidate is not None and len(self.candidate) > 0 and self.subset is not None and len(self.subset) > 0:
                print("pose")
                self.candidate=None
                self.subset=None
            if len(self.faces_locations) > 0:
                for name,face_location in zip(self.names,self.faces_locations):
                    y1, x2, y2, x1 = face_location
                    cv2.rectangle(self.canvas, (x1, y1), (x2, y2), (0, 0, 255), 3)
                    cv2.putText(self.canvas,name, (x1, y1 - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.9, (0, 0, 255), 2)
                self.names=[]
                self.faces_locations=[]
            cv2.imshow('demo', self.frame)
            if cv2.waitKey(30) & 0xFF == ord('q'):
                break
        cv2.destroyAllWindows()


utils=ImageUtils()
